packages/convex-helpers/server/zod4.ts

```diff
@@ -0,0 +1,642 @@
+  TableName extends TableNamesInDataModel<DataModel> = TableNamesInDataModel<DataModel>,
+>(tableName: TableName) {
+  const base = z
+    .string()
+    .refine((s) => typeof s === "string" && s.length > 0, {
+      message: `Invalid ID for table "${tableName}"`,
+    })
+    .transform((s) => s as GenericId<TableName>)
+    .brand(`ConvexId_${tableName}`)
+    .describe(`convexId:${tableName}`);
+  registryHelpers.setMetadata(base, { isConvexId: true, tableName });
+  return base as z.ZodType<GenericId<TableName>>;
+}
+
+export const withSystemFields = <
+  Table extends string,
+  T extends { [key: string]: z.ZodTypeAny },
+>(tableName: Table, zObject: T) => {
+  return { ...zObject, _id: zid(tableName as any), _creationTime: z.number() } as const;
+};
+
+function isZid(schema: z.ZodTypeAny): boolean {
+  const m = registryHelpers.getMetadata(schema);
+  return !!(m && m.isConvexId && typeof m.tableName === "string");
+}
+
+function makeUnion(members: GenericValidator[]) {
+  const arr = members.filter((m): m is GenericValidator => !!m);
+  if (arr.length === 0) return v.any();
+  if (arr.length === 1) return arr[0];
+  const [a, b, ...rest] = arr as [GenericValidator, GenericValidator, ...GenericValidator[]];
+  return v.union(a, b, ...rest);
+}
+
+// Narrow helpers for Convex validators
+function asValidator(x: unknown): any {
+  return x as unknown as any;
+}
+
+// Zod v4 -> Convex (input)
+function zodToConvexInternal(
+  zodValidator: z.ZodTypeAny,
+  visited: Set<z.ZodTypeAny> = new Set(),
+): GenericValidator {
+  if (!zodValidator) return v.any();
+  if (visited.has(zodValidator)) return v.any();
+  visited.add(zodValidator);
+
+  if (isZid(zodValidator)) {
+    const meta = registryHelpers.getMetadata(zodValidator);
+    return v.id((meta?.tableName as string) ?? "unknown");
+  }
+
+  if (zodValidator instanceof z.ZodDefault) {
+    const inner = zodValidator.unwrap() as unknown as z.ZodTypeAny;
+    const innerV = zodToConvexInternal(inner, visited);
+    return v.optional(asValidator(innerV));
+  }
+  if (zodValidator instanceof z.ZodOptional) {
+    const inner = zodValidator.unwrap() as unknown as z.ZodTypeAny;
+    const innerV = zodToConvexInternal(inner, visited);
+    return v.optional(asValidator(innerV));
+  }
+  if (zodValidator instanceof z.ZodNullable) {
+    const inner = zodValidator.unwrap() as unknown as z.ZodTypeAny;
+    const innerV = zodToConvexInternal(inner, visited);
+    return v.union(asValidator(innerV), v.null());
+  }
+
+  if (zodValidator instanceof z.ZodString) return v.string();
+  if (zodValidator instanceof z.ZodNumber) return v.float64();
+  if (zodValidator instanceof z.ZodBigInt) return v.int64();
+  if (zodValidator instanceof z.ZodBoolean) return v.boolean();
+  if (zodValidator instanceof z.ZodNull) return v.null();
+  if (zodValidator instanceof z.ZodAny) return v.any();
+  if (zodValidator instanceof z.ZodUnknown) return v.any();
+  if (zodValidator instanceof z.ZodDate) return v.float64();
```

Comment:
This is incorrect, Convex’s `v.float64` doesn’t accept `Date` objects.

Maybe we want to give a default handling of dates with a codec instead?